### PR TITLE
core: fix restoring chain with `StateRootInHeader = true`

### LINKED ISSUE #3597: Restoring chain with `StateRootInHeader = true` can fail
# Problem:
`db restore` results depend on the db state. The state root stored for genesis block is not valid if service was started with non-empty database containing a single genesis block.

# To reproduce
The test to reproduce (put it to `pkg/core`).
BTW, this is the easiest way I have found to share storage between chains, could you tell me if this can be done easier?
```
type nopCloserStorage struct {
	storage.Store
}

func (nopCloserStorage) Close() error {
	return nil
}

func TestBlockchainRestoreStateRootInHeader(t *testing.T) {
	bc := newTestChainWithCustomCfg(t, func(c *config.Config) {
		c.ProtocolConfiguration.StateRootInHeader = true
	})
	require.NoError(t, bc.AddBlock(bc.newBlock()))

	w := io.NewBufBinWriter()
	require.NoError(t, chaindump.Dump(bc, w.BinWriter, 0, 2))
	require.NoError(t, w.Err)

	data := w.Bytes()
	fcfg := func(c *config.Config) { c.ProtocolConfiguration.StateRootInHeader = true }
	initChain := func(st storage.Store) *Blockchain {
		chain := initTestChain(t, st, fcfg)
		go chain.Run()
		return chain
	}

	t.Run("this passes", func(t *testing.T) {
		bc := initChain(storage.NewMemoryStore())
		r := io.NewBinReaderFromBuf(data)
		require.NoError(t, chaindump.Restore(bc, r, 0, 2, nil))
		bc.Close()
	})
	t.Run("this doesn't pass", func(t *testing.T) {
		st := nopCloserStorage{Store: storage.NewMemoryStore()}
		{
			bc := initChain(st)
			r := io.NewBinReaderFromBuf(data)
			require.NoError(t, chaindump.Restore(bc, r, 0, 1, nil))
			bc.Close()
		}
		{
			bc := initChain(st)
			r := io.NewBinReaderFromBuf(data)
			require.NoError(t, chaindump.Restore(bc, r, 1, 1, nil))
			bc.Close()
		}
	})
}
```

Results:
```
$ go test -count=1 -v .
...
=== NAME  TestBlockchainRestoreStateRootInHeader/this_doesn't_pass
    blockchain_core_test.go:185:
                Error Trace:    /repo/neo-go/pkg/core/blockchain_core_test.go:185
                Error:          Received unexpected error:
                                failed to add block 1: state root for previous block is invalid: 6fc66e00ccd8ed7ae0563e9d2f2b859c2abbede741fb546ceb99e5253e8aa41d != 0000000000000000000000000000000000000000000000000000000000000000
                Test:           TestBlockchainRestoreStateRootInHeader/this_doesn't_pass
=== NAME  TestBlockchainRestoreStateRootInHeader
    logger.go:146: 2024-10-04T17:24:54.486+0300 INFO    persisted to disk       {"blocks": 1, "keys": 141, "headerHeight": 1, "blockHeight": 1, "took": "27.268µs"}
--- FAIL: TestBlockchainRestoreStateRootInHeader (0.01s)
    --- PASS: TestBlockchainRestoreStateRootInHeader/this_passes (0.00s)
    --- FAIL: TestBlockchainRestoreStateRootInHeader/this_doesn't_pass (0.00s)
FAIL
```

I have quickly glanced upon tests for chaindump and it seems we never close the blockchain in those tests, just combine multiple `Restore()` calls.

### PR DESCRIPTION
**Pull request by AliceInHunterland**
Close #3597

### CODE REVIEW

File: pkg/core/stateroot/module.go [hunk 1]
```diff
@@ -183,9 +183,16 @@ func (s *Module) Init(height uint32) error {
 		updateStateHeightMetric(h)
 	}
 
-	if height == 0 {
+	if s.mpt == nil {
 		s.mpt = mpt.NewTrie(nil, s.mode, s.Store)
+	}
+
+	if height == 0 {
 		s.currentLocal.Store(util.Uint256{})
```
Comment (roman-khimov) [inline pkg/core/stateroot/module.go:191]: Maybe move this into `err != nil` branch?

File: pkg/core/stateroot/module.go [hunk 2]
```diff
@@ -183,9 +183,16 @@ func (s *Module) Init(height uint32) error {
...
+		sr, err := s.GetStateRoot(height)
+		if err == nil {
+			s.currentLocal.Store(sr.Root)
+		}
 		return nil
 	}
 	r, err := s.getStateRoot(makeStateRootKey(height))
```
Comment (roman-khimov) [inline pkg/core/stateroot/module.go:198]: And maybe this can be reused.

**Review by roman-khimov — COMMENTED** [review]
At least it's not _that_ bad and can be fixed easily.

File: pkg/core/blockchain_core_test.go [hunk 3]
```diff
@@ -414,3 +416,73 @@ func TestNewBlockchain_InitHardforks(t *testing.T) {
...
+		bc.Close()
+
+		bc = initChain(st)
+		actual := bc.stateRoot.CurrentLocalStateRoot()
+		require.Equal(t, expected, actual)
+		r = io.NewBinReaderFromBuf(data)
+		require.NoError(t, chaindump.Restore(bc, r, 1, 1, nil))
+		bc.Close()
+	})
+
```
Comment (AnnaShaleva) [inline pkg/core/blockchain_core_test.go:470]: Useless line.

**Review by AnnaShaleva — COMMENTED** [review]
Makes sense.

File: pkg/core/blockchain_core_test.go (hunk 3 above)
Comment (AliceInHunterland) [reply pkg/core/blockchain_core_test.go]: fixed.

**Review by AnnaShaleva — APPROVED** [review]

**Pull request merged**